src/mono/wasm/debugger/BrowserDebugProxy/EvaluateExpression.cs

```diff
@@ -247,75 +249,123 @@ private static string ConvertJSToCSharpLocalVariableAssignment(string idName, JT
-                if (value == null)
-                    throw new ReturnAsErrorException($"Failed to resolve member access for {memberAccessString}", "ReferenceError");
-
-                memberAccessValues.Add(value);
-            }
-
-            return memberAccessValues;
+            IList<JObject> values = new List<JObject>();
+            foreach (T element in collectionToResolve)
+                values.Add(await resolutionFunc(element, resolver, token));
+            return values;
         }
 
-        private static async Task<IList<JObject>> ResolveIdentifiers(IEnumerable<IdentifierNameSyntax> identifiers, MemberReferenceResolver resolver, CancellationToken token)
+        private static async Task<JObject> ResolveMemberAccessExpression(MemberAccessExpressionSyntax member_access,
+                                MemberReferenceResolver resolver, CancellationToken token)
         {
-            var values = new List<JObject>();
-            foreach (IdentifierNameSyntax var in identifiers)
-            {
-                JObject value = await resolver.Resolve(var.Identifier.Text, token);
-                if (value == null)
-                    throw new ReturnAsErrorException($"The name {var.Identifier.Text} does not exist in the current context", "ReferenceError");
-
-                values.Add(value);
-            }
+            string memberAccessString = member_access.ToString();
+            JObject value = await resolver.Resolve(memberAccessString, token);
+            return value ?? throw new ReturnAsErrorException($"Failed to resolve member access for {memberAccessString}", "ReferenceError");
+        }
 
-            return values;
+        private static async Task<JObject> ResolveIdentifier(IdentifierNameSyntax identifier,
+                                MemberReferenceResolver resolver, CancellationToken token)
+        {
+            JObject value = await resolver.Resolve(identifier.Identifier.Text, token);
+            return value ?? throw new ReturnAsErrorException($"The name {identifier.Identifier.Text} does not exist in the current context", "ReferenceError");
         }
 
-        private static async Task<IList<JObject>> ResolveMethodCalls(IEnumerable<InvocationExpressionSyntax> methodCalls, Dictionary<string, JObject> memberAccessValues, MemberReferenceResolver resolver, CancellationToken token)
+        private static async Task<(IList<JObject>, IList<JObject>, IList<JObject>)> ResolveMethodCalls(ExpressionSyntaxReplacer replacer, MemberReferenceResolver resolver, CancellationToken token)
         {
-            var values = new List<JObject>();
-            foreach (InvocationExpressionSyntax methodCall in methodCalls)
+            var methodCallValues = new List<JObject>();
+            // used for replacing method call on primitive:
+            var maesValues = new List<JObject>();
+            var identifierValues = new List<JObject>();
```

Comment:
The maximum number of objects that these can have is == `replacer.methodCalls.Count`, so we can use `new List<JObject>(capacity: numCalls);`. This will make sure that the lists don't need to resize their internal arrays.